src/lib/docker-utils.ts

```diff
@@ -72,37 +64,18 @@ export async function fetchDeltaWithProgress(
 	onProgress: ProgressCallback,
 	serviceName: string,
 ): Promise<string> {
-	const deltaSourceId =
-		deltaOpts.deltaSourceId != null
-			? deltaOpts.deltaSourceId
-			: deltaOpts.deltaSource;
-
-	const timeout = deltaOpts.deltaApplyTimeout;
-
 	const logFn = (str: string) =>
 		log.debug(`delta([${serviceName}] ${deltaOpts.deltaSource}): ${str}`);
 
-	if (!_.includes([2, 3], deltaOpts.deltaVersion)) {
+	if (deltaOpts.deltaVersion !== 3) {
 		logFn(
 			`Unsupported delta version: ${deltaOpts.deltaVersion}. Falling back to regular pull`,
 		);
 		return await fetchImageWithProgress(imgDest, deltaOpts, onProgress);
 	}
 
-	// We need to make sure that we're not trying to apply a
-	// v3 delta on top of a v2 delta, as this will cause the
-	// update to fail, and we must fall back to a standard
-	// image pull
-	if (
-		deltaOpts.deltaVersion === 3 &&
-		(await isV2DeltaImage(deltaOpts.deltaSourceId))
-	) {
-		logFn(`Cannot create a delta from V2 to V3, falling back to regular pull`);
-		return await fetchImageWithProgress(imgDest, deltaOpts, onProgress);
```

Comment:
I think we’ll need to keep this check.

The image on a device that has been updating using rsync deltas lacks the metadata needed to make it a “proper” docker image, which means trying to use it as a source to apply a v3 delta over fails. The only way forward in this case is to perform a regular pull — effectively performing a one-time migration by pulling the complete new image (and subsequently removing the old one).

The reason we still need this check even though the update path to balenaOS 3 will involve an update to the latest balenaOS versions first (hence recent Supervisors that know how to migrate rsync deltas) is that unless the user *also* pushes a new release in between, the local images may still be in that incompatible image format. We can technically only drop this check when all devices on the platform have switched away from rsync deltas.

Might worth expanding the code comment a bit to include this context.

Reply:
@dfunckt Apologies for dropping the ball on this. I've re-added `isV2DeltaImage` and amended the comment above `if (deltaVersion === 3 && isV2DeltaImage)` to reflect the context you've provided. Also re-added tests for this util and for the v3-on-v2-delta case,